Cleanup, saturation and breathe

jamadagni opened the pull request:
Cleanups technically breaks API in the case of the brightness and update delay but makes for cleaner code and the fix on user side is minor.

Also exposes the saturation to the user which can be useful if they want more pastel colors.

The breathe addition depends on https://github.com/keyboardio/Kaleidoscope-LEDControl/pull/27 as it uses the new `breath_compute_helper` to avoid extra call to `millis()` for each LED in case of the wave effect.

jamadagni committed:
expose hue_steps, saturation and other cleanups

hue_steps control useful mainly to change direction of rainbow
saturation control useful to get more pastel vs vibrant colours
no point in trivial setters/getters [this slightly breaks API :-(]
simplify variable names
remove misleading comments re hue range
reduce hue variable type to 8-bits so we don't need to enforce 255 max value
make hue_steps variable type signed to enable changing direction of rainbow

jamadagni committed:
add breathe option

note that the variable update delay has been renamed since if breathing is on,
LED updates aren't delayed and this is only the interval between hue updates

jamadagni committed:
documentation update corresponding to previous changes

obra commented:
In general, we try hard not to break user code without a really, really good reason. And when we do, we need to document it extensively and offer compile time explanations of the change and how to fix calling code.  

Also, I’d prefer one logical change per PR. Added features and cleaning up code should each get their own pull request. 
I’d be happy to look at a revised set of PRs.

jamadagni commented:
OK I understand your point and willing to do so, but when one set of changes applies above another, how to make one PR depend on another?

jamadagni commented:
> In general, we try hard not to break user code without a really, really good reason. And when we do, we need to document it extensively and offer compile time explanations of the change and how to fix calling code.

The problem here is that the settings like brightness and update delay should not have been made private in the first place. The settings started getting exposed at ea90ca09b045e041dc07dc6357588504bd7eca84, first readonly and then read-write access was given. Really it is useful to expose all "settings" and only the *state* ie current hue and last hue update time should be kept private.

Note that for each setting, keeping the variable private entails a pair of accessors and multiple lines of code both in the header and source, but really those do no validation or anything to justify their separate existence and merely add to the clutter.

Switching to the new API means just changing a function call to an assignment for the setter which is trivial to do. But how to add a compile time warning/advice for that I'm not sure.

Only in the case of the hue update delay, because the addition of breathing means no delays, the function name could not be retained as the variable name. The only way to retain would be to move the breathing version out to a separate plugin which again means duplicated code.

Kindly apprise as to how you want to take this forward. Personally I think the user community is small enough that an announcement of API changes on the forums is sufficient, especially since the change would only be to a few lines of client code.

(Note that I could always just retain the changes on my system for personal use, but am submitting the PR only in trying to contribute back. ☺)

obra commented:
I'm traveling with family this week, so I can't give you the detailed
explanation this deserves, but in general we prefer getters and setters
over assignment to variables because it gives is a lot more flexibility to
update the internal implementation. When we started out, we definitely had
the same opinion you did, opting for the simplest internal implementation.
Over time, we've come to realize that making things easier and safer and
clearer for end-users is far more important than our own convenience.

They're not clutter so much as 'good encapsulation' -- it's unfortunate
that we weren't diligent about this from the earliest days. The compiler
takes care of optimizing everything out, so there's no runtime overhead.

Despite the style difference, I do appreciate the pull request :)
ᐧ

On Fri, Aug 3, 2018 at 5:37 AM Shriramana Sharma <notifications@github.com>
wrote:

> In general, we try hard not to break user code without a really, really
> good reason. And when we do, we need to document it extensively and offer
> compile time explanations of the change and how to fix calling code.
>
> The problem here is that the settings like brightness and update delay
> should not have been made private in the first place. The settings started
> getting exposed at ea90ca0
> <https://github.com/keyboardio/Kaleidoscope-LEDEffect-Rainbow/commit/ea90ca09b045e041dc07dc6357588504bd7eca84>,
> first readonly and then read-write access was given. Really it is useful to
> expose all "settings" and only the *state* ie current hue and last hue
> update time should be kept private.
>
> Note that for each setting, keeping the variable private entails a pair of
> accessors and multiple lines of code both in the header and source, but
> really those do no validation or anything to justify their separate
> existence and merely add to the clutter.
>
> Switching to the new API means just changing a function call to an
> assignment for the setter which is trivial to do. But how to add a compile
> time warning/advice for that I'm not sure.
>
> Only in the case of the hue update delay, because the addition of
> breathing means no delays, the function name could not be retained as the
> variable name. The only way to retain would be to move the breathing
> version out to a separate plugin which again means duplicated code.
>
> Kindly apprise as to how you want to take this forward. Personally I think
> the user community is small enough that an announcement of API changes on
> the forums is sufficient, especially since the change would only be to a
> few lines of client code.
>
> (Note that I could always just retain the changes on my system for
> personal use, but am submitting the PR only in trying to contribute back. ☺)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Kaleidoscope-LEDEffect-Rainbow/pull/10#issuecomment-410240548>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACxaNAL6GzT83CgGCY6HCOz8CMrUIATks5uNEQTgaJpZM4VruJL>
> .
>

jamadagni committed:
Limit refresh rate and don't update on every cycle even if breathing

The breathe function is somewhat costly and is found to cause drag in
mouse movements. This commit seeks to fix this problem.

It is observed that the function doesn't change output value for every
input value. It only causes the output brightness to increase by 128
units (from 80 to 208) over 2048 ms (the half-period). This means 1 unit
for 16 ms. But a brightness change of 1 unit doesn't mean much visually
especially considering persistence of vision. A refresh rate of 20 per
second ie 50 ms between LED updates is found to be sufficient to avoid
the drag effect while maintaining smoothness in brightness changes.
